kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/DomainMetadata.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.actions;
+
+import static io.delta.kernel.internal.util.InternalUtils.requireNonNull;
+import static java.util.Objects.requireNonNull;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.data.Row;
+import io.delta.kernel.internal.data.GenericRow;
+import io.delta.kernel.types.BooleanType;
+import io.delta.kernel.types.StringType;
+import io.delta.kernel.types.StructType;
+import java.util.HashMap;
+import java.util.Map;
+
+/** Delta log action representing an `DomainMetadata` action */
+public class DomainMetadata {
+  /** Full schema of the {@link DomainMetadata} action in the Delta Log. */
+  public static final StructType FULL_SCHEMA =
+      new StructType()
+          .add("domain", StringType.STRING, false /* nullable */)
+          .add("configuration", StringType.STRING, false /* nullable */)
+          .add("removed", BooleanType.BOOLEAN, false /* nullable */);
+
+  public static DomainMetadata fromColumnVector(ColumnVector vector, int rowId) {
+    if (vector.isNullAt(rowId)) {
+      return null;
+    }
+    return new DomainMetadata(
+        requireNonNull(vector.getChild(0), rowId, "domain").getString(rowId),
+        requireNonNull(vector.getChild(1), rowId, "configuration").getString(rowId),
+        requireNonNull(vector.getChild(2), rowId, "removed").getBoolean(rowId));
+  }
+
+  public static DomainMetadata fromRow(Row row) {
```

Comment:
resurrecting a comment from below so it doesn't get lost: we should probably document exceptions for this right?

ref: https://github.com/delta-io/delta/pull/3835/files/b27ca3e672f94581645e78a66822fb3e851d081d#r1852969975

Reply:
Yeah we should. I've added a doc for this and a better error message.